### PR TITLE
`path` and `assocPath` to support negative indexes

### PR DESCRIPTION
**Pull request by keithbro**
As required for #2093.

### CODE REVIEW

File: src/assocPath.js [hunk 1]
```diff
@@ -41,7 +43,8 @@ module.exports = _curry3(function assocPath(path, val, obj) {
   }
   if (_isInteger(idx) && _isArray(obj)) {
     var arr = [].concat(obj);
-    arr[idx] = val;
+    var arrIdx = idx < 0 ? arr.length + idx : idx;
```
Comment (keithbro) [inline src/assocPath.js]: I'm wondering if it would be nicer if this logic was handled by `assoc`?

File: test/assocPath.js [hunk 2]
```diff
@@ -49,4 +49,9 @@ describe('assocPath', function() {
     eq(assocPath(['foo', 'bar', 'baz'], 42, {foo: null}), {foo: {bar: {baz: 42}}});
   });
 
+  it('supports array indices', function() {
+    eq(assocPath([0], 42, [1, 2, 3]), [42, 2, 3]);
+    eq(assocPath([-1], 42, [1, 2, 3]), [1, 2, 42]);
```
Comment (keithbro) [inline test/assocPath.js]: it's not clear to me what the expected result of 

    assocPath([99], 42, [1, 2, 3]);

should be?

File: src/assocPath.js (hunk 1 above)
Comment (kedashoe) [inline src/assocPath.js]: I think it is the other way around? `assoc` can be easily defined in terms of `assocPath`. They have historically had divergent behavior with certain input, which is why this hasn't happened yet. Maybe it is time we do that, as this PR as it stands would be yet another difference.

File: src/assocPath.js [hunk 3]
```diff
@@ -29,6 +29,8 @@ var isNil = require('./isNil');
  *
  *      // Any missing or non-object keys in path will be overridden
  *      R.assocPath(['a', 'b', 'c'], 42, {a: 5}); //=> {a: {b: {c: 42}}}
+ *
+ *      R.assocPath([-1], 42, [1,2,3]); //=> [1,2,42]
  */
 module.exports = _curry3(function assocPath(path, val, obj) {
   if (path.length === 0) {
```
Comment (kedashoe) [inline src/assocPath.js]: github won't let me comment on a line that isn't part of the diff, but I think we need some logic here to handle multiple negative indexes (eg `assocPath([-1, -1], 10, [[1, 2, 3], [4, 5, 6]])`):

https://github.com/keithbro/ramda/blob/d8bb8c71e70ca26d74aec1ffeafd4e6eabc80b17/src/assocPath.js#L40-L43

let's add a test for this as well

File: test/assocPath.js (hunk 2 above)
Comment (kedashoe) [inline test/assocPath.js]: I assume whatever js would normally do here is fine.. create a sparse array?

File: src/assocPath.js (hunk 1 above)
Comment (keithbro) [inline src/assocPath.js]: I'm not sure. I feel like the logic of associating a value to an index of a flat array or a property given an flat object is sufficiently complex to warrant it's own function (`assoc`). `assocPath` could use that function when `path.length === 1`. If we put all the logic in `assocPath` then I think we are going to end up with quite a complex function. But perhaps I'm missing a simple solution. 

I'll push up a solution I think is good and you can let me know what you think.

File: test/assocPath.js (hunk 2 above)
Comment (buzzdecafe) [inline test/assocPath.js]: > I assume whatever js would normally do here is fine.. create a sparse array?

😱

Comment (keithbro) [inline test/assocPath.js]: For positive indices, sparse makes sense.. but for negative it gets confusing. In Perl it throws and error which seems sane:

    $ perl -e '[]->[-1] = 1'
    Modification of non-creatable array value attempted, subscript -1 at -e line 1.

File: src/assocPath.js (hunk 1 above)
Comment (kedashoe) [inline src/assocPath.js]: Sorry for the delay on this one @keithbro . I'd prefer `assoc` to be `function assoc(prop, val, obj) { return assocPath([prop], val, obj); }` and for all the logic to be in `assocPath`. But let's see what others think.. anyone?

File: src/assoc.js [hunk 4]
```diff
@@ -22,6 +23,16 @@ var _curry3 = require('./internal/_curry3');
  *      R.assoc('c', 3, {a: 1, b: 2}); //=> {a: 1, b: 2, c: 3}
  */
 module.exports = _curry3(function assoc(prop, val, obj) {
+  if (_isArray(obj)) {
+    var newArr = [].concat(obj);
+    if (prop < 0 && Math.abs(prop) > newArr.length) {
+      throw new Error('Invalid index');
```
Comment (PopeFelix) [inline src/assoc.js]: You might change this to "index out of range"

Comment (davidchambers) [inline src/assoc.js]: Also, this should be a [RangeError][1].


[1]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RangeError

**Comment by CrossEye** [comment]
While there is definitely some work to do in figuring out the relationships between our `prop` and `path` functions and in our handling of negative indices, this has fallen behind some other work.  We'll address the underlying issue post `v1.0`.

**Pull request closed without merging**